### PR TITLE
Refactor the list parsing script to use a dedicated parser library

### PR DESCRIPTION
**Pull request by englehardt**
I've replicated the parsing logic in an [external library](https://github.com/mozilla/trackingprotection-tools/blob/4a1c0e1aea588e952264f5d0a98acdc075d36050/trackingprotection_tools/DisconnectParser.py). This PR switches `lists2safebrowsing` over to the external library.

This has a few clear benefits:
* I need to run analyses using the Disconnect list, so I've had to replicate the `lists2safebrowsing` logic. Prior to this I've had to effectively maintain two slightly different implementations with each structural change.
* The new parser has [fairly comprehensive test suite](https://github.com/mozilla/trackingprotection-tools/blob/4a1c0e1aea588e952264f5d0a98acdc075d36050/trackingprotection_tools/test/test_disconnect_parser.py), which will hopefully take a lot of the guess work and manual testing out of new structural changes to the list.
* The new structure allows us the flexibility to specify compound filters in the list creation script, i.e., "domains in (categoryA or categoryB) and in categoryC", I'm going to do this in an upcoming PR.

I recommended first reviewing and merging #71. We can rebase this PR to make the diffs much more readable.

This parser matches the output of the current script when run on `prod.ini`, with the exception of a minor bug in the current script. I can't do an exact hash comparison of the log outputs like I did in #71, since I removed some of the print statements from the original script. Instead, we can compare the missing domains file-by-file using the log output:

There are no domains in new files that aren't in the old files. You can verify this by running the current list creation script, copying the output to `original/`, running the script in this PR, and then running `ls -1 *.log | xargs -n 1 -I {} bash -c "echo {} & diff <(cat original/{} | sort) <(cat {} | sort) | grep '^>' | grep '\[m\]' | sort | uniq "`.

Following the same procedure as above we can look at the domains in the old lists that aren't in the new list (via `ls -1 *.log | xargs -n 1 -I {} bash -c "echo {} & diff <(cat original/{} | sort) <(cat {} | sort) | grep '^<' | grep '\[m\]' | sort | uniq "`):
```
adguard-ads-digest256.log
ads-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] vcmedia.vn >> vcmedia.vn/
allow-flashallow-digest256.log
analytics-track-digest256.log
base-cryptomining-track-digest256.log
baseeff-track-digest256.log
base-fingerprinting-track-digest256.log
base-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] rkdms.com >> rkdms.com/
< [m] trackingsoft.com >> trackingsoft.com/
< [m] vcmedia.vn >> vcmedia.vn/
basew3c-track-digest256.log
block-flash-digest256.log
block-flashsubdoc-digest256.log
content-cryptomining-track-digest256.log
contenteff-track-digest256.log
content-fingerprinting-track-digest256.log
content-track-digest256.log
contentw3c-track-digest256.log
easylist-ads-digest256.log
easyprivacy-ads-digest256.log
except-flashallow-digest256.log
except-flash-digest256.log
except-flashinfobar-digest256.log
except-flashsubdoc-digest256.log
fanboyannoyance-ads-digest256.log
fanboysocial-ads-digest256.log
fastblock1-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] trackingsoft.com >> trackingsoft.com/
< [m] vcmedia.vn >> vcmedia.vn/
fastblock1-trackwhite-digest256.log
fastblock2-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] trackingsoft.com >> trackingsoft.com/
< [m] vcmedia.vn >> vcmedia.vn/
fastblock2-trackwhite-digest256.log
fastblock3-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] trackingsoft.com >> trackingsoft.com/
< [m] vcmedia.vn >> vcmedia.vn/
mozfullstaging-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] rkdms.com >> rkdms.com/
< [m] trackingsoft.com >> trackingsoft.com/
< [m] vcmedia.vn >> vcmedia.vn/
mozfull-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] rkdms.com >> rkdms.com/
< [m] trackingsoft.com >> trackingsoft.com/
< [m] vcmedia.vn >> vcmedia.vn/
mozplugin2-block-digest256.log
mozplugin-block-digest256.log
mozstdstaging-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] rkdms.com >> rkdms.com/
< [m] trackingsoft.com >> trackingsoft.com/
< [m] vcmedia.vn >> vcmedia.vn/
mozstdstaging-trackwhite-digest256.log
mozstd-track-digest256.log
< [m] adonnetwork.com >> adonnetwork.com/
< [m] buzzcity.com >> buzzcity.com/
< [m] i-behavior.com >> i-behavior.com/
< [m] matomy.com >> matomy.com/
< [m] rkdms.com >> rkdms.com/
< [m] trackingsoft.com >> trackingsoft.com/
< [m] vcmedia.vn >> vcmedia.vn/
mozstd-trackwhite-digest256.log
social-track-digest256.log
```

Here we see that several files are missing the same set of domains. These domains are all in the list under multiple categories and/or organizations, and are actually added to the output lists multiple times in the current list creation script (which appears to be a bug based on [this line](https://github.com/mozilla-services/shavar-list-creation/blob/7291d5f7af861c11aef38848492a0f735032f6e0/lists2safebrowsing.py#L245-L246)). Our current script only adds them once, hence the difference. We should file follow-up bugs with Disconnect to address these duplicates.

### CODE REVIEW

**Comment by englehardt** [comment]
> We should file follow-up bugs with Disconnect to address these duplicates.

FYI These will be addressed in https://github.com/mozilla-services/shavar-prod-lists/pull/56#issuecomment-471753178.

**Comment by englehardt** [comment]
Fixes #61

**Comment by englehardt** [comment]
This has been rebased on top of #71

**Comment by groovecoder** [comment]
Oh duh ... I ran my spot-check before I finished reading the PR description. I see that the domain differences are because "Our current script only adds them once, hence the difference."

File: lists2safebrowsing.py [hunk 1]
```diff
@@ -216,25 +226,62 @@ def add_domain_to_list(domain, previous_domains, allow_list, log_file, output):
...
+def get_domains_from_filters(parser, which_dnt, list_categories, desired_tags):
+    """Apply filters to the Disconnect list to return a set of matching domains
 
     Args:
-      blocklist_json: A JSON blob containing Disconnect's list.
-      allow_list: Hosts that we can't put on the blocklist.
-      chunk: The chunk number to use.
-      output_file: A file-handle to the output file.
+      parser : A DisconnectParser instance
       log_file: A filehandle to the log file.
```
[the remark this answers is not in the document]
Comment (englehardt) [reply lists2safebrowsing.py]: This will be fixed in #73 so I'll merge without fixing

**Comment by englehardt** [comment]
Thanks!

**Pull request merged**